### PR TITLE
[Internal] UnitTest: Adds Test to Validate Regions Type Public Fields

### PR DESCRIPTION
**Pull request by kirankumarkolli**
[INTERNAL] UnitTest: Regions type public fields validation

Broken for (Germany Central)  and (Germany Northeast). 
Needs Direct package upgrade with the fix.

### CODE REVIEW

**Review by kundadebdatta — APPROVED** [review]
LGTM!

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/CosmosClientOptionsUnitTests.cs [hunk 1]
```diff
@@ -718,6 +719,32 @@ public void VerifyApplicationRegionSettingsWithPreferredRegions()
...
+
+        private static IEnumerable<object[]> GetTestData()
+        {
+            List<object[]> testData = new List<object[]>();
+            foreach (FieldInfo fieldInfo in typeof(Regions).GetFields().Where(e => e.IsPublic && e.IsStatic))
+            {
+                string regionValue = fieldInfo.GetValue(null).ToString();
+
+                // TODO: 
+                if (!regionValue.Equals("Germany Central", StringComparison.OrdinalIgnoreCase)
```
Comment (kundadebdatta) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/CosmosClientOptionsUnitTests.cs:738]: nit: Can be inferred from `Regions.GermanyCentral` and `Regions.GermanyNorthEast`.

**Review by kundadebdatta — DISMISSED** [review]
LGTM.

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/CosmosClientOptionsUnitTests.cs [hunk 2]
```diff
@@ -718,6 +719,32 @@ public void VerifyApplicationRegionSettingsWithPreferredRegions()
...
+        }
+
+        private static IEnumerable<object[]> GetTestData()
+        {
+            List<object[]> testData = new List<object[]>();
+            foreach (FieldInfo fieldInfo in typeof(Regions).GetFields().Where(e => e.IsPublic && e.IsStatic))
+            {
+                string regionValue = fieldInfo.GetValue(null).ToString();
+
+                // TODO:
```
Comment (philipthomas-MSFT) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/CosmosClientOptionsUnitTests.cs:737]: what is the TODO?

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/CosmosClientOptionsUnitTests.cs [hunk 3]
```diff
@@ -718,6 +719,32 @@ public void VerifyApplicationRegionSettingsWithPreferredRegions()
...
+
+        private static IEnumerable<object[]> GetTestData()
+        {
+            List<object[]> testData = new List<object[]>();
+            foreach (FieldInfo fieldInfo in typeof(Regions).GetFields().Where(e => e.IsPublic && e.IsStatic))
+            {
+                string regionValue = fieldInfo.GetValue(null).ToString();
+
+                // TODO: 
+                if (!regionValue.Equals(Regions.GermanyCentral, StringComparison.OrdinalIgnoreCase)
```
Comment (philipthomas-MSFT) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/CosmosClientOptionsUnitTests.cs:738]: So, are we testing other Regions in other tests?

**Review by philipthomas-MSFT — DISMISSED** [review]

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/CosmosClientOptionsUnitTests.cs (hunk 3 above)
Comment (kirankumarkolli) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/CosmosClientOptionsUnitTests.cs:738]: These two regions are deprecated and removed hence excluding them

**Review by philipthomas-MSFT — DISMISSED** [review]

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/CosmosClientOptionsUnitTests.cs [hunk 4]
```diff
@@ -718,6 +719,31 @@ public void VerifyApplicationRegionSettingsWithPreferredRegions()
             cosmosClientOptions.GetConnectionPolicy(clientId: 0);
         }
 
+        [TestMethod]
+        [DynamicData(nameof(GetPublicRegionNames), DynamicDataSourceType.Method)]
+        public void VerifyApplicationRegionSettingsForAllPublicRegions(string regionName)
+        {
+            CosmosClientOptions cosmosClientOptions = new CosmosClientOptions { ApplicationRegion = regionName };
+            cosmosClientOptions.GetConnectionPolicy(clientId: 0);
```
Comment (Pilchie) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/CosmosClientOptionsUnitTests.cs:727]: Should we be asserting something about the returned ConnectionPolicy, or we're just happy that it doesn't throw?

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/CosmosClientOptionsUnitTests.cs [hunk 5]
```diff
@@ -718,6 +719,31 @@ public void VerifyApplicationRegionSettingsWithPreferredRegions()
...
+        public void VerifyApplicationRegionSettingsForAllPublicRegions(string regionName)
+        {
+            CosmosClientOptions cosmosClientOptions = new CosmosClientOptions { ApplicationRegion = regionName };
+            cosmosClientOptions.GetConnectionPolicy(clientId: 0);
+        }
+
+        private static IEnumerable<object[]> GetPublicRegionNames()
+        {
+            List<object[]> regionNames = new List<object[]>();
+            foreach (FieldInfo fieldInfo in typeof(Regions).GetFields().Where(e => e.IsPublic && e.IsStatic))
```
Comment (Pilchie) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/CosmosClientOptionsUnitTests.cs:733]:
```suggestion
            foreach (FieldInfo fieldInfo in typeof(Regions).GetFields(BindingFlags.Public & BindingFlags.Static))

```

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/CosmosClientOptionsUnitTests.cs (hunk 4 above)
Comment (kirankumarkolli) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/CosmosClientOptionsUnitTests.cs:727]: It throw exception in-case of unsupported regions.

Comment (Pilchie) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/CosmosClientOptionsUnitTests.cs:727]: Nit: Maybe a comment saying that we're just verifying that it doesn't throw?

**Review by Pilchie — DISMISSED** [review]

**Review by Pilchie — DISMISSED** [review]

**Pull request merged**